common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -508,6 +524,302 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+
+            val resetPasswordSubmitApiResult = performResetPasswordSubmitCall(
+                oAuth2Strategy = oAuth2Strategy,
+                parameters = parameters
+            )
+
+            return when (resetPasswordSubmitApiResult) {
+                is ResetPasswordSubmitApiResult.SubmitSuccess -> {
+                    resetPasswordPollCompletion(
+                        oAuth2Strategy = oAuth2Strategy,
+                        passwordResetToken = resetPasswordSubmitApiResult.passwordResetToken,
+                        pollIntervalInSeconds = resetPasswordSubmitApiResult.pollInterval
+                    )
+                }
+                is ResetPasswordSubmitApiResult.PasswordInvalid -> {
+                    ResetPasswordCommandResult.PasswordNotAccepted(
+                        error = resetPasswordSubmitApiResult.error,
+                        errorDescription = resetPasswordSubmitApiResult.errorDescription
+                    )
+                }
+
+                is ResetPasswordSubmitApiResult.ExpiredToken,
+                is ResetPasswordSubmitApiResult.UnknownError -> {
+                    Logger.warn(
+                        TAG,
+                        "Unexpected result: $resetPasswordSubmitApiResult"
+                    )
+                    resetPasswordSubmitApiResult as ApiErrorResult
+                    INativeAuthCommandResult.UnknownError(
+                        error = resetPasswordSubmitApiResult.error,
+                        errorDescription = resetPasswordSubmitApiResult.errorDescription,
+                        details = resetPasswordSubmitApiResult.details
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in resetPasswordSubmitNewPassword", e)
+            throw e
+        }
+    }
+
+    private fun resetPasswordPollCompletion(
+        oAuth2Strategy: NativeAuthOAuth2Strategy,
+        passwordResetToken: String,
+        pollIntervalInSeconds: Int?
+    ): ResetPasswordSubmitNewPasswordCommandResult {
+        fun pollCompletionTimedOut(startTime: Long): Boolean {
+            val currentTime = System.currentTimeMillis()
+            return currentTime - startTime > ResetPasswordSubmitNewPasswordCommand.POLL_COMPLETION_TIMEOUT_IN_MILISECONDS
+        }
+
+        fun pollIntervalIsAppropriate(pollIntervalInSeconds: Int?): Boolean {
+            return pollIntervalInSeconds != null && pollIntervalInSeconds <= 15 && pollIntervalInSeconds >= 1
+        }
+
+        val methodTag = "$TAG:resetPasswordPollCompletion"
+
+        LogSession.logMethodCall(TAG, "${TAG}.resetPasswordPollCompletion")
+
+        try {
+            val pollWaitInterval: Int = if (!pollIntervalIsAppropriate(pollIntervalInSeconds)) {
+                ResetPasswordSubmitNewPasswordCommand.DEFAULT_POLL_COMPLETION_INTERVAL_IN_MILISECONDS
+            } else {
+                pollIntervalInSeconds!! * 1000
```

Comment:
Why is the double exclamation needed? Aren't Ints non-null by default in Kotlin?

Also why are we multiplying by 1000 here?

Reply:
`pollIntervalInSeconds` is typed as `Int?` and that can be null. We are multiplying by 1000 to convert seconds into milliseconds.

Reply:
Can we use a type of Int that is NonNull?

Reply:
I have changed it.